### ISSUE TITLE
Create new Cancer Biomarkers summary widget and detail view for evidence page

### ISSUE THREAD

**Issue by ireneisdoomed** [opentargets/platform#1735]
We are adding Cancer Biomarkers as a new evidence source for the upcoming release (21.09). We need to create a table to display the data.

<details>
<summary>The current schema is in a mature state and no changes are expected:</summary>

```python
root
 |-- biomarkerName: string (nullable = true)
 |-- biomarkers: struct (nullable = true)
 |    |-- geneExpression: array (nullable = true)
 |    |    |-- element: struct (containsNull = true)
 |    |    |    |-- id: string (nullable = true)
 |    |    |    |-- name: string (nullable = true)
 |    |-- variant: array (nullable = true)
 |    |    |-- element: struct (containsNull = true)
 |    |    |    |-- functionalConsequenceId: string (nullable = true)
 |    |    |    |-- id: string (nullable = true)
 |    |    |    |-- name: string (nullable = true)
 |-- confidence: string (nullable = true)
 |-- datasourceId: string (nullable = true)
 |-- datatypeId: string (nullable = true)
 |-- diseaseFromSource: string (nullable = true)
 |-- diseaseFromSourceMappedId: string (nullable = true)
 |-- drugFromSource: string (nullable = true)
 |-- drugId: string (nullable = true)
 |-- drugResponse: string (nullable = true)
 |-- literature: array (nullable = true)
 |    |-- element: string (containsNull = true)
 |-- targetFromSourceId: string (nullable = true)
 |-- urls: array (nullable = true)
 |    |-- element: struct (containsNull = true)
 |    |    |-- niceName: string (nullable = true)
 |    |    |-- url: string (nullable = true)
```
</details>

Evidence strings can be found here: `gs://otar000-evidence_input/CancerBiomarkers/json`, although they are not yet exposed in the GraphQL API.

The data at the moment has a bug that prevents it from passing validation (see #1705), but it shouldn't be a blocker to scope how data should look like.

The data represents how tumour alterations can modulate drug response. We have different scenarios:

<details>
<summary>1. An example where biomarker describes a single variant (and we have variantId)</summary>

```python
{
    "datasourceId": "cancer_genome_interpreter",
    "datatypeId": "affected_pathway",
    "drugFromSource": "ERK inhibitors",
    "drugResponse": "GO_0042493",
    "targetFromSourceId": "BRAF",
    "diseaseFromSource": "Head and neck squamous cell carcinoma",
    "diseaseFromSourceMappedId": "EFO_0000181",
    "confidence": "Case report",
    "biomarkerName": "BRAF (G469A)",
    "urls": [
        {
            "niceName": "American Society of Clinical Oncology Annual Meeting 2017",
            "url": "https://meetinglibrary.asco.org/browse-meetings/2017%20ASCO%20Annual%20Meeting"
        }
    ],
    "biomarkers": {
        "variant": [
            {
                "name": "BRAF:G469A",
                "id": "7_140481402_C_G",
                "functionalConsequenceId": "SO_0001777"
            }
        ]
    }
}
```
</details>

<details>
<summary>2. An example where biomarker describes multiple variants</summary>

```python
{
    "datasourceId": "cancer_genome_interpreter",
    "datatypeId": "affected_pathway",
    "drugFromSource": "Pd1 Ab Inhibitors",
    "targetFromSourceId": "KRAS",
    "diseaseFromSource": "Lung adenocarcinoma",
    "diseaseFromSourceMappedId": "EFO_0000571",
    "confidence": "Early trials",
    "biomarkerName": "STK11 oncogenic mutation + KRAS oncogenic mutation",
    "urls": [
        {
            "niceName": "American Society of Clinical Oncology Annual Meeting 2017",
            "url": "https://meetinglibrary.asco.org/browse-meetings/2017%20ASCO%20Annual%20Meeting"
        }
    ],
    "biomarkers": {
        "variant": [
            {
                "name": "KRAS",
                "functionalConsequenceId": "SO_0001777"
            },
            {
                "name": "STK11",
                "functionalConsequenceId": "SO_0001777"
            }
        ]
    }
}
```
</details>

<details>
<summary>3. An example where biomarker consists of a change in expression</summary>

```python
{
    "datasourceId": "cancer_genome_interpreter",
    "datatypeId": "affected_pathway",
    "drugFromSource": "Ado-trastuzumab Emtansine",
    "drugResponse": "GO_0042493",
    "targetFromSourceId": "ERBB2",
    "diseaseFromSource": "Breast adenocarcinoma",
    "diseaseFromSourceMappedId": "EFO_0000304",
    "confidence": "FDA guidelines",
    "biomarkerName": "ERBB2 overexpression",
    "urls": [
        {
            "niceName": "Food and Drug Administration guidelines",
            "url": "https://www.fda.gov/RegulatoryInformation/Guidances/default.htm"
        }
    ],
    "biomarkers": {
        "geneExpression": [
            {
                "name": "ERBB2:over",
                "id": "GO_0010628"
            }
        ]
    }
}
```
</details>

<details>
<summary>4. An example where biomarker describes a variant and a change in expression </summary>

```python
{
    "datasourceId": "cancer_genome_interpreter",
    "datatypeId": "affected_pathway",
    "drugFromSource": "Trastuzumab",
    "drugId": "CHEMBL1201585",
    "targetFromSourceId": "ANXA1",
    "diseaseFromSource": "Breast adenocarcinoma",
    "diseaseFromSourceMappedId": "EFO_0000304",
    "confidence": "Early trials",
    "biomarkerName": "ARID1A amplification + ANXA1 overexpression",
    "literature": [
        "27172896"
    ],
    "biomarkers": {
        "variant": [
            {
                "name": "ARID1A:amp",
                "functionalConsequenceId": "SO_0001563"
            }
        ],
        "geneExpression": [
            {
                "name": "ANXA1:over",
                "id": "GO_0010628"
            }
        ]
    }
}
```
</details>

Things to bear in mind:
1. No evidence strings have `biomarkerName` set to null.
2. ~~There are cases where `drugResponse` is null - this will be fixed as soon as EFO creates new terms for certain drug responses.~~ This is no longer the case, as we are displaying the raw string until we are happy with how EFO treats drug responses.
3. No evidence strings have `drugFromSource` set to null.
4. `drugFromSource` can be mapped sometimes to a CHEMBL drug (`drugId`)
5. `biomarkers.variant.functionalConsequenceId` is a SO ID that needs to be resolved. We want to show the label and build the respective link.
6. Similarly, `biomarkers.geneExpression.id` is a GO ID that needs to be resolved. We want to show the label and build the respective link.
7. Most of the evidence is supported by a PubMed ID, sometimes the reference is built into `urls` displaying. Note that we might have cases where both fields are present (`biomarkerName='ERBB2 overexpression'`)

**Comment by andrewhercules** [comment]
Based on the analyses above by @ireneisdoomed, I have created a sample API query and draft design specification for a new Cancer Biomarkers section on the evidence page.

Please note that both are subject to further change and iteration.

### Query

```
query targetInfo {
  target(ensemblId: "ENSG00000157764") {
    id
    approvedSymbol
    evidences(efoIds:["EFO_0000181"], datasourceIds:["cancer_biomarkers"]){
      rows{
        disease{
          id
          name
        }
        diseaseFromSource
        biomarkerName
        biomarkers{
          geneExpression{
            id
            name
          }
          variant{
            functionalConsequenceId
            id
            name
          }          
        }
        drugFromSource
        drugId 
        drugResponse
        confidence
        literature
        urls {
          niceName
          url
        }
      }
    }
  }
}
```

### Summary widget

<img width="275" alt="Screenshot 2021-09-02 at 01 18 03" src="https://user-images.githubusercontent.com/7490258/131761882-79ba8786-e316-485b-a0cc-70c74ba77c9a.png">

- [ ] Display the number of rows returned by the API query.

### Detail view

- [ ] Create a data table section based on the following [design specification](https://docs.google.com/presentation/d/1hRwGMYGrMwhxKchQkuBiEzgXb868w6LhGRgqcYhBftE)

![Cancer Biomarkers (1)](https://user-images.githubusercontent.com/7490258/131763451-2f2c1fe0-9d25-4176-9a54-6079ad61c8c9.png)

- [ ] For the section description, display the following text.

> Expert-curated genomic biomarkers of drug sensitivity, resistance, and toxicity for drugs targeting **GENE** with an experimental or approved indication of **DISEASE**. Source: [Cancer Genome Interpreter](https://www.cancergenomeinterpreter.org/2018/biomarkers).

- [ ] Show the "Affected pathway" datatype chip in the top right-hand corner

- [ ] Add the ability to search the table by the "Disease", "Biomarker", "Reported drug", and "Source" columns

- [ ] Add the "Show more rows" functionality
 
- [ ] In the "Disease" column, display the `disease.name` value and link to the relevant disease profile page using the `disease.id` value. Also, display the `diseaseFromSource` in a tooltip.

- [ ] In the "Biomarker" column, display the `biomarkerName` value and on-click, open the drawer component. Iterate through the `biomarkers` object and construct up to two unordered lists — one list called "Variant" with the `biomarkers.variant.name` and `biomarkers.variant.id` (if available); the other list called "Gene expression" with the `biomarkers.geneExpression.name`. It is important to note that there are associations with only one gene expression, one variant, multiple variants, and both gene expression and variant information — see @ireneisdoomed analyses above for examples. And also note that in further iterations, this data will be expanded and the SO and GO terms will be resolvable entities.

![Cancer Biomarkers (3)](https://user-images.githubusercontent.com/7490258/131763558-8137e8df-cf8f-4cb1-959a-4c6d48c9ccfa.png)

- [ ] In the "Reported drug" column, display the `drugFromSource` value and if `drugId` is not `null`, link to the drug profile page using the `drugId` value — otherwise, just display the `drugFromSource` value.

- [ ] In the "Source" column, display the `confidence` value

- [ ] In the "Literature" column, use both the `literature` and `urls` arrays and display the PubMed IDs and `url.niceName` values in the drawer component and link to Europe PMC. Note: if the drawer component is flexible enough to show the citation/abstract and other URLs, we can use that functionality — otherwise, we can just show a list of the IDs and `url.NiceName` values.

**Comment by andrewhercules** [comment]
@mirandaio I've updated the design specification above — it should be good to go! 👍

**Comment by ireneisdoomed** [comment]
Thank you @mirandaio for your work!

Some things I've noted:

1. There are special characters in the name of the data source in the facet. Could we change that to `Cancer Biomarkers`?
<img width="318" alt="image" src="https://user-images.githubusercontent.com/45119610/142704293-97205472-b13f-416e-a042-123574ca9af1.png">

2. I don't know if it is intended, but the order of the columns are changed. Could we put the `Drug response` column immediately after `Reported drug`?
<img width="1250" alt="image" src="https://user-images.githubusercontent.com/45119610/142704447-beadeeec-c41e-427a-8293-aedba0198a8a.png">

3. There are biomarkers which cannot be referenced with a Pubmed ID, hence lacking any type of link out on the FE. For this cases an url is also returned in the API (`urls.niceName` and `urls.url`) instead. ([link](https://main--platform-app.netlify.app/evidence/ENSG00000196712/EFO_0000658))
<img width="1250" alt="image" src="https://user-images.githubusercontent.com/45119610/142705347-97fa1864-4555-4323-aa92-1a591e25357f.png">
@andrewhercules, do you think it would be a good idea to collapse the `Source` and `Literature` columns similarly as we do in the target safety widget in such a way that:
- references where `literature != null` display the literature widget
- references where `literature == null` link out to the url returned in `urls.url`?


4. The biomarker name can be very long and look ugly in the table. `ABL1` is the most extreme case, with about 70 mutations. ([link](https://main--platform-app.netlify.app/evidence/ENSG00000097007/EFO_0000339))
<img width="1250" alt="image" src="https://user-images.githubusercontent.com/45119610/142705904-856f8f89-c1b0-4b1a-a953-1dafb1555935.png">
@andrewhercules , what is the best way to go about this? Is it possible to collapse the name and adding `...`, e.g.? Please note that the second most promiscuous case is found on `NOTCH1` (152 characters), which doesn't look bad ([link](https://main--platform-app.netlify.app/evidence/ENSG00000148400/EFO_0000311)):
<img width="1250" alt="image" src="https://user-images.githubusercontent.com/45119610/142706494-01c61d4a-4889-4a68-ae12-f5e994b2a547.png">

**Comment by andrewhercules** [comment]
@mirandaio based on the feedback from @ireneisdoomed above, can we please do the following?

- [ ] Change the order of the columns and move the "Drug response" column left so that it is after the "Reported drug" column

- [ ] In the "Biomarker" column, truncate the text to show the first 100 characters and then show ellipsis and on-hover, show the full text

- [ ] Update the "Source" column to use the drawer component and iterate through the `urls` array and show a list of each `urls.niceName` and link with the `urls.url` value

- [ ] Update the association page facets — see #1736

**Comment by mirandaio** [comment]
@ireneisdoomed @andrewhercules I have addressed the feedback. Let me know if it's ready to merge. Thanks!

**Comment by andrewhercules** [comment]
Thanks @mirandaio — looks great! 👍 

@ireneisdoomed we had to adjust how we handle the `urls` and for now, we are showing in the `Source` column. This allows us to handle situations where there's both a publication and a trial ID or external link (e.g. [ALK and lymphoma](https://main--platform-app.netlify.app/evidence/ENSG00000171094/EFO_0000574))

**Comment by d0choa** [comment]
Niceeee

**Comment by ireneisdoomed** [comment]
Thank you both!

@andrewhercules To me it looks great now. The distinction of the origin of the data is clearer like this.
@mirandaio, I don't know if it is due to this specific PR, but it seems that something has been changed on the widgets drawer. I don't think it was like this on Friday. Do you see the same behaviour? ([link](https://main--platform-app.netlify.app/evidence/ENSG00000068078/EFO_0000292))

<img width="1367" alt="image" src="https://user-images.githubusercontent.com/45119610/142996128-3212aae1-b8f0-4df0-98ba-860f60876325.png">

**Comment by andrewhercules** [comment]
Ticket closed as work completed, merged into `main`, and released in `21.11`
